### PR TITLE
re-add majority voting

### PR DESCRIPTION
**Pull request by evanmcc**
reverting the old unanimous voting code.

### CODE REVIEW

File: src/transactions/v1/blockchain_txn_vars_v1.erl [hunk 1]
```diff
@@ -419,21 +419,24 @@ validate_master_keys(Txn, Gen, Artifact, Ledger) ->
...
                     KeyProofs =
                         case multi_key_proofs(Txn) of
                             [] ->
                                 throw({error, no_multi_keys_proofs});
                             Ps ->
                                 Ps
                         end,
-                    [case verify_key(Artifact, Key, KeyProof) of
+                    Votes = blockchain_utils:count_votes(Artifact, MultiKeys, KeyProofs),
+                    case Votes >= length(ProofKeys) of
```
Comment (Vagabond) [inline src/transactions/v1/blockchain_txn_vars_v1.erl:432]: What are we actually checking for here, maybe add some comments?

Comment (evanmcc) [reply src/transactions/v1/blockchain_txn_vars_v1.erl:432]: will do.

File: src/transactions/v1/blockchain_txn_vars_v1.erl [hunk 2]
```diff
@@ -279,13 +279,13 @@ is_valid(Txn, Chain) ->
                                 %% handle the case where this gets set before
                                 %% the keys are set
                                 case blockchain_ledger_v1:multi_keys(Ledger) of
-                                    {ok, MultiKeys} ->
+                                    {ok, MultiKeys} when MultiKeys /= [] ->
```
[the remark this answers is not in the document]
Comment (madninja) [reply src/transactions/v1/blockchain_txn_vars_v1.erl]: ah ok... however the behavior still is that if there are _no_ multi keys we'd fall back to the master_key?

File: src/transactions/v1/blockchain_txn_vars_v1.erl [hunk 3]
```diff
@@ -419,21 +419,37 @@ validate_master_keys(Txn, Gen, Artifact, Ledger) ->
...
+                    %% keys, and to ensure that all new keys have one proof associated with them.
+                    {ok, OldMultiKeys} = blockchain_ledger_v1:multi_keys(Ledger),
+                    %% remove all existing keys from the new list. They don't need to re-prove
+                    %% themselves and cannot 'vote' for the change here.  Their votes as to the
+                    %% signedness of the transaction are counted elsewhere.
+                    ProofKeys = MultiKeys -- OldMultiKeys,
                     KeyProofs =
-                        case multi_key_proofs(Txn) of
+                        %% deduplicate the proofs
+                        case lists:usort(multi_key_proofs(Txn)) of
```
Comment (Vagabond) [inline src/transactions/v1/blockchain_txn_vars_v1.erl:434]: Would they be duplicated in normal operation somehow, or is this in case of an attack?

Comment (Vagabond) [reply src/transactions/v1/blockchain_txn_vars_v1.erl:434]: ECDSA does not give deterministic signatures, so deduping will not remove repeated signatures by the same key. EdDSA *is* deterministic happily, so this is not a concern with ed25519 keys.

**Review by Vagabond — APPROVED** [review]

File: src/blockchain_utils.erl [hunk 4]
```diff
@@ -464,8 +471,8 @@ poc_per_hop_max_witnesses(Ledger) ->
     end.
 
 
-%% majority(N) ->
-%%     N div 2 + 1.
+majority(N) ->
+    N div 2 + 1.
```
Comment (vihu) [inline src/blockchain_utils.erl:475]:
```suggestion
    (N div 2) + 1.
```

File: src/transactions/v1/blockchain_txn_vars_v1.erl (hunk 2 above)
[the remark this answers is not in the document]
Comment (evanmcc) [reply src/transactions/v1/blockchain_txn_vars_v1.erl]: yes

**Review by Vagabond — APPROVED** [review]

**Review by vihu — APPROVED** [review]

**Pull request merged**